File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    Set<Integer> numbers = new HashSet<>();
    
    public int solution(String numbersStr) {
        // dfs 돌려서 경우의 수 set 저장
        // 소수인것만 골라서 계산
        int n = numbersStr.length();
        boolean[] visited = new boolean[n];
        dfs("", visited, numbersStr, n);
        int result = 0;
        
        for (int num: numbers){
            if (isPrime(num)) result ++;
        }
        return result;
    }
    
    void dfs(String current, boolean[] visited, String numbersStr, int n){
        if (current != ""){
            numbers.add(Integer.parseInt(current));
        }
        
        for (int i = 0; i < n; i++){
            if (!visited[i]){
                visited[i] = true;
                dfs(current + numbersStr.charAt(i), visited, numbersStr, n);
                visited[i] = false;
            }
        }
        return;
    }

    boolean isPrime(int x){
        if (x < 2) return false;
        
        for (int i = 2; i * i <= x; i ++){
            if (x % i == 0) return false;
        }
        return true;
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
def solution(numbers_str):
    numbers = set()
    n = len(numbers_str)
    visited = [False] * n
    
    def is_prime(x):
        if x < 2:
            return False
        for i in range(2, int(x ** 0.5) + 1):
            if x % i == 0:
                return False
        return True
    
    def dfs(current):
        if current:
            numbers.add(int(current))  
        for i in range(n):
            if not visited[i]:
                visited[i] = True
                dfs(current + numbers_str[i])
                visited[i] = False
    
    dfs("")
    return sum(1 for num in numbers if is_prime(num))